### PR TITLE
Disable severity combobox for non-configurable rules

### PR DESCRIPTION
**Pull request by Youssef1313**
Small part of https://github.com/dotnet/roslyn/issues/53171.

image

### CODE REVIEW

**Review by mavasani — CHANGES_REQUESTED** [review]
I think this will be confusing. IMO we should show the descriptors but explicitly disable severity changes in the UI by greying out and disabling the combo box.

@jmarolf for his view.

**Comment by Youssef1313** [comment]
> I think this will be confusing. IMO we should show the descriptors but explicitly disable severity changes in the UI by greying out and disabling the combo box.
> 
> @jmarolf for his view.

I thought the whole point of EditorConfig UI is to allow easily editing editorconfig files without having to remember IDs and option names/values or looking them up in the docs, etc.

I don't see the value of having a greyed out row. Why would I care about "NotConfigurable" rules if I'm using a feature made to configure rules?

To avoid the confusion, we might add a note at the top saying that it's a list of configurable rules.

**Comment by mavasani** [comment]
> I thought the whole point of EditorConfig UI is to allow easily editing editorconfig files 

The goal of the UX is to allow easy viewing, searching and editing _all_ the applicable analyzers rules. Not seeing non-configurable rules in this UX would lead one to believe their analyzer package is missing these rules completely, which are listed in Analyzers node as well as analyzer package documentation. Additionally, it will be unclear why they are unable to configure or suppress a diagnostic for such a rule. A disabled entry in the UX with a hover message about non-configurability makes this explicit.

**Review by mavasani — APPROVED** [review]

File: src/EditorFeatures/Core/EditorConfigSettings/Data/AnalyzerSetting.cs [hunk 1]
```diff
@@ -44,6 +46,7 @@ internal class AnalyzerSetting
         public DiagnosticSeverity Severity { get; private set; }
         public bool IsEnabled { get; private set; }
         public Language Language { get; }
+        public bool IsNotConfigurable { get; set; }
```
Comment (Youssef1313) [inline src/EditorFeatures/Core/EditorConfigSettings/Data/AnalyzerSetting.cs:49]:
```suggestion
        public bool IsNotConfigurable { get; }
```

**Comment by Youssef1313** [comment]
> Publishing build artifacts failed with an error: Not found PathtoPublish: D:\workspace\_work\1\s\artifacts\bin\Microsoft.VisualStudio.LanguageServices.IntegrationTests\Debug\net472\xUnitResults

Closing and reopening.

**Comment by sharwell** [comment]
@Youssef1313 Can you rebase this onto release/dev16.11?

**Review by jmarolf — APPROVED** [review]

**Comment by jmarolf** [comment]
Thanks so much @Youssef1313 really like this change!

**Comment by sharwell** [comment]
> Don't show descriptors with NotConfigurable tag in EditorConfig UI

Note that this is still the desirable behavior for rules that are both non-configurable and have default severity Hidden. The current form of this pull request fixes cases where the non-configurable rule has severity Info or higher, but we will need to go back and filter out the Hidden rules entirely.

**Comment by mavasani** [comment]
> we will need to go back and filter out the Hidden rules entirely.

Why? That seems very weird. What is the harm in showing these hidden non-configurable rules in UX?

**Comment by sharwell** [comment]
> Why? That seems very weird. What is the harm in showing these hidden non-configurable rules in UX?

All such cases I'm aware of, including the two cases in #53171, two additional cases from Roslyn in the past, and cases in StyleCop Analyzers, are descriptors created _with the direct intent_ to hide the diagnostic from the UI.

**Comment by Youssef1313** [comment]
I rebased into dev16.11 branch.

I'm not against any of the two approaches, however, the approach I prefer more is to not show them at all ( what I initially had in https://github.com/dotnet/roslyn/commit/17a80900467d9079c015f949f185993f48f17a04). @mavasani @sharwell @jmarolf I'd be happy to revert back to the first approach or keep the PR as-is once you discuss this and decide the way it should be.

**Comment by mavasani** [comment]
> I'd be happy to revert back to the first approach

Please don't - as I stated above, searching applicable analyzers is one of the biggest feature requirements for this UX. It is not just about editing. Ruleset editor was mostly popular because of the ability to search all applicable rules - editing the values can easily be done manually as well once you get used to the syntax.

**Comment by mavasani** [comment]
> descriptors created with the direct intent to hide the diagnostic from the UI

These analyzers are wrongly authored. If you want a non-configurable hidden diagnostic, it should be authored as a refactoring. If you are doing so just to change the applicable span for code fix, again the issue here is our system does not support specifying this additional metadata about diagnostics/code fix presentation.

I stand by my view that a P0 for this UX is ability to search for every single applicable analyzer rule ID. IMO, its even more important feature then editing the severity/code style values. Adding filters such as hiding non-configurable hidden diagnostics, may be fine as a short term workaround (though I am still not sure if it is required). It is absolutely not the right long term solution.

**Comment by sharwell** [comment]
> Ruleset editor was mostly popular because of the ability to search all applicable rules ...

The rule set editor could not search all applicable rules. It followed the rule I have been requesting: all rules _except_ ones which are both non-configurable and have hidden severity.

> These analyzers are wrongly authored.

The analyzers are authored according to patterns established in Visual Studio 2015 (Roslyn 1.x) and still supported. Other parts of the UI established and maintained precedent that authors can reliably expect these diagnostics to not appear in the UI. There are two places, both in newer UIs, where these currently appear but need to be removed.

> Adding filters such as hiding non-configurable hidden diagnostics, may be fine as a short term workaround (though I am still not sure if it is required). It is absolutely not the right long term solution.

It is required and expected. Third-party analyzers have been designed and implemented according to this well-established pattern, and there is no open request for it to change as it relates to any feature. For all examples provided so far, there is no value in showing the diagnostic. If this behavior is not blocking any feature development, why would we care to change it?

**Comment by JoeRobich** [comment]
Changing target branch to 16.11 servicing branch

**Comment by jmarolf** [comment]
retargeting to dev17 as this missed dev16.11

**Comment by jmarolf** [comment]
@Youssef1313 unless you have an objections I plan to merge this.

**Comment by Youssef1313** [comment]
> @Youssef1313 unless you have an objections I plan to merge this.

I don't. Let's get it in 🎉

**Pull request merged**